In [150]:
# Initialize Otter
import otter
grader = otter.Notebook("hw03.ipynb")

<img src="./ccsf.png" alt="CCSF Logo" width=200px style="margin:0px -5px">

# Homework 03: Tables

---

## Textbook References

* [Sequences](https://inferentialthinking.com/chapters/05/Sequences.html)
* [Tables](https://inferentialthinking.com/chapters/06/Tables.html)

## Assignment Reminders

- 🚨 Make sure to run the code cell at the top of this notebook that starts with `# Initialize Otter` to load the auto-grader.
- Your Tasks are categorized as auto-graded (📍) and manually graded (📍🔎).
    - For all the auto-graded tasks:
        - Replace the `...` in the provided code cell with your own code.
        - Run the `grader.check` code cell to run some tests on your code.
        - Keep in mind that for homework and project assignments, sometimes there are hidden tests that you will not be able to see the results of that we use for scoring the correctness of your response. **Passing the auto-grader does not guarantee that your answer is correct.**
    - For all the manually graded tasks:
        - You might need to provide your own response to the provided prompt. Do so by replacing the template text "_Type your answer here, replacing this text._" with your own words.
        - You might need to produce a graphic or something else using code. Do so by replacing the `...` in the code cell to generate the image, table, etc.
        - In either case, review the rubric on the associated <a href="https://ccsf.instructure.com" target="_blank">Canvas</a> Assignment page to understand the scoring criteria.
- Throughout this assignment and all future ones, please be sure to not re-assign variables throughout the notebook! _For example, if you use `max_temperature` in your answer to one question, do not reassign it later on. Otherwise, you will fail tests that you thought you were passing previously!_
- You may [submit](#Submit-Your-Assignment-to-Canvas) this assignment as many times as you want before the deadline. Your instructor will score the last version you submit once the deadline has passed.
- We encourage you to discuss this assignment with others but make sure to write and submit your own code. Refer to the syllabus to learn more about how to learn cooperatively.

---

## Configure the Notebook

Run the following cell to configure this Notebook.

In [152]:
import numpy as np
from datascience import *
import matplotlib
%matplotlib inline
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

---

## Fruit

To begin this homework, you are going to practice with some of the `Table` methods you learned about this week.

### Task 01 📍

Suppose you have a small fruit stand that has an inventory of 4 apples, 3 oranges, and 1 pineapple.

Create a table called `fruits` that contains this information:
* It should have two columns: `'fruit name'` and `'amount'`.
* It should have three rows, one for each type of fruit.
* The values for `'fruit name'` should be the strings `'apple'`, `'orange'`, and `'pineapple'`
* The values for `'amount'` should be integers `4`, `3`, and `1`.

*Note:* Use lower-case and singular words for the name of each fruit, like `"apple"`.

_Points:_ 2

In [25]:
fruits = Table().with_columns(
    'fruit name', make_array('apple', 'orange', 'pineapple'),
    'amount', make_array(4, 3, 1)
)

In [26]:
grader.check("task_01")

task_01 results: All test cases passed!
task_01 - 1 message: ✅ fruits is a Table.
task_01 - 2 message: ✅ fruits looks correct!

---

### Task 02 📍

The file `inventory.csv` contains information about the inventory at a fruit stand.  Each row represents the contents of one box of fruit. Load it as a table named `inventory` using the `Table.read_table()` function. 

**Note**: `Table.read_table(...)` takes one argument (data file name in string format) and returns a table.

_Points:_ 2

In [27]:
inventory = Table.read_table("inventory.csv")
inventory

box ID | fruit name | count
53686  | kiwi       | 45
57181  | strawberry | 123
25274  | apple      | 20
48800  | orange     | 35
26187  | strawberry | 255
57930  | grape      | 517
52357  | strawberry | 102
43566  | peach      | 40

In [28]:
grader.check("task_02")

task_02 results: All test cases passed!
task_02 - 1 message: ✅ inventory is a Table.
task_02 - 2 message: ✅ inventory looks correct.

---

### Task 03 📍

Does each box at the fruit stand contain a different fruit? Set `all_different` to `True` if each box contains a different fruit or to `False` if multiple boxes contain the same fruit.

*Hint:* You don't have to write code to calculate the True/False value for `all_different`. Just look at the `inventory` table and assign `all_different` to either `True` or `False` according to what you can see from the table in answering the question.

_Points:_ 2

In [29]:
all_different = True
all_different

True

In [30]:
grader.check("task_03")

task_03 results: All test cases passed!
task_03 - 1 message: ✅ all_different is assigned to True or False.

---

### Task 04 📍

The file `sales.csv` contains the number of fruit sold from each box last Saturday.  It has an extra column called `price per fruit ($)` that's the price *per item of fruit* for fruit in that box.  The rows are in the same order as the `inventory` table.  Load these data into a table called `sales`.

_Points:_ 2

In [50]:
sales = Table().read_table('sales.csv')
sales

box ID | fruit name | count sold | price per fruit ($)
53686  | kiwi       | 3          | 0.5
57181  | strawberry | 101        | 0.2
25274  | apple      | 0          | 0.8
48800  | orange     | 35         | 0.6
26187  | strawberry | 25         | 0.15
57930  | grape      | 355        | 0.06
52357  | strawberry | 102        | 0.25
43566  | peach      | 17         | 0.8

In [51]:
grader.check("task_04")

task_04 results: All test cases passed!
task_04 - 1 message: ✅ sales is a Table.
task_04 - 2 message: ✅ sales looks correct.

---

### Task 05 📍

How many fruits did the store sell in total on that day, last Saturday?

_Points:_ 2

In [66]:
total_fruits_sold = sum(sales.column('count sold'))
total_fruits_sold

638

In [67]:
grader.check("task_05")

task_05 results: All test cases passed!
task_05 - 1 message: ✅ total_fruits_sold is an integer.
task_05 - 2 message: ✅ total_fruits_sold is a possible quantity.

---

### Task 06 📍

What was the store's total revenue (the total price of all fruits sold) on that day? Assign that value to `total_revenue`.

*Hint:* If you're stuck, think first about how you would compute the total revenue from just the grape sales. It should be `355 * 0.06`, which gives `21.3`.

_Points:_ 2

In [68]:
total_revenue = sum(sales.column('count sold')*sales.column('price per fruit ($)'))
total_revenue

106.84999999999999

In [69]:
grader.check("task_06")

task_06 results: All test cases passed!
task_06 - 1 message: ✅ total_revenue is a possible value.
task_06 - 2 message: ✅ total_revenue doesn't just reflect the revenue for grapes.

---

### Task 07 📍

Make a new table called `remaining_inventory`.  It should have the same rows and columns as `inventory`, except that the amount of fruit sold from each box should be subtracted from that box's **original** count, so that the `count` column is **updated to be** the amount of fruit remaining after Saturday.

**Notes**:
* As a check, since `355` grapes were sold from the original `517`, then `162` grapes should be in the remaining inventory once you're done.
* We recommend that you do not do the subtractions individually. Let Python (with the help of arrays) do the hard work for you!
* Don't forget that you can always add extra lines of code and intermediate variables. The auto-grader is only checking the value of `remaining_inventory`.

_Points:_ 3

In [85]:
remaining_values = inventory.column('count') - sales.column('count sold')
remaining_inventory = inventory.with_columns('count', remaining_values)
remaining_inventory

box ID | fruit name | count
53686  | kiwi       | 42
57181  | strawberry | 22
25274  | apple      | 20
48800  | orange     | 0
26187  | strawberry | 230
57930  | grape      | 162
52357  | strawberry | 0
43566  | peach      | 23

In [86]:
grader.check("task_07")

task_07 results: All test cases passed!
task_07 - 1 message: ✅ remaining_inventory is a Table.
task_07 - 2 message: ✅ The updated grape inventory looks good.

---

## Unemployment


The [Federal Reserve Bank of St. Louis](https://fred.stlouisfed.org/categories/33509) publishes data about jobs in the US.  Below, we've loaded data on unemployment in the United States. There are many ways of defining unemployment, and our dataset includes two notions of the unemployment rate:

1. Among people who are able to work and are looking for a full-time job, the percentage who can't find a job.  This is called the Non-Employment Index, or NEI.
2. Among people who are able to work and are looking for a full-time job, the percentage who can't find any job *or* are only working at a part-time job.  The latter group is called "Part-Time for Economic Reasons", so the acronym for this index is NEI+PTER.  (Economists are great at marketing.)

The data that you will use in this assignment contains [quarterly average NEI percentages from 1994 until 2024](https://fred.stlouisfed.org/series/NEIM156SFRBRIC) and [quarterly average NEI+PTER percentages from 1994 until 2024](https://fred.stlouisfed.org/series/NEIPTERM156SFRBRIC)

*In this assignment, you will do some preliminary analysis of these unemployment metrics. In a future assignment, you will work on creating visualizations and summarizing your analysis.*


---

### Task 08 📍

The data are in a CSV file called `unemployment.csv`.  Load the data in that file into a `Table` and assign the table to the name `unemployment`.


_Points:_ 1

In [87]:
unemployment = Table().read_table('unemployment.csv')
unemployment

DATE       | NEI     | NEI+PTER
1994-01-01 | 9.82912 | 10.9749
1994-04-01 | 9.42543 | 10.6306
1994-07-01 | 9.22842 | 10.359
1994-10-01 | 8.98376 | 10.1199
1995-01-01 | 8.88245 | 10.0086
1995-04-01 | 8.97931 | 10.0942
1995-07-01 | 8.96056 | 10.1052
1995-10-01 | 8.95893 | 10.0813
1996-01-01 | 8.90291 | 9.97698
1996-04-01 | 8.80475 | 9.88223
... (110 rows omitted)

In [88]:
grader.check("task_08")

task_08 results: All test cases passed!
task_08 - 1 message: ✅ It seems that you have assigned the name unemployment to a table.
task_08 - 2 message: ✅ The contents in unemployment seem correct.

---

### Task 09 📍

Dates can be tricky to work with. When you loaded the data into the table `unemployment`, the dates were read as strings. Run the following cell to see this is true.

In [89]:
type(unemployment.column('DATE').item(0))

str

Strings do not always make the best sense to use for dates since we organize dates by months, years, etc. and this is different than the alphabetical order we typically use for strings. A common date type for dates is [`datetime`](https://docs.python.org/3/library/datetime.html). In this class, you are not responsible for knowing how to work directly with the `datetime` format. Instead, we will set up the code to help you with that date type.

**Run the following code cell to convert all the dates in the `'DATE'` column to `date` format.**

_Points:_ 0

In [90]:
# Just run this cell
import datetime
from datetime import datetime, date

def update_date_format(date_string):
    date_object = datetime.strptime(date_string, '%Y-%m-%d').date()
    return date_object

if not isinstance(unemployment.column('DATE').item(0),  date):
    unemployment = unemployment.with_column(
        'DATE', unemployment.apply(update_date_format, 'DATE'))
    print("The date type for the values in the 'DATE' column have been updated.\n")
else:
    print("The date type is correct for the values in the 'DATE' column.\n")
    
display(unemployment)

The date type for the values in the 'DATE' column have been updated.



DATE       | NEI     | NEI+PTER
1994-01-01 | 9.82912 | 10.9749
1994-04-01 | 9.42543 | 10.6306
1994-07-01 | 9.22842 | 10.359
1994-10-01 | 8.98376 | 10.1199
1995-01-01 | 8.88245 | 10.0086
1995-04-01 | 8.97931 | 10.0942
1995-07-01 | 8.96056 | 10.1052
1995-10-01 | 8.95893 | 10.0813
1996-01-01 | 8.90291 | 9.97698
1996-04-01 | 8.80475 | 9.88223
... (110 rows omitted)

In [91]:
grader.check("task_09")

task_09 results: All test cases passed!
task_09 - 1 message: ✅ You've converted all the dates.

---

### Task 10 📍

For this task:
1. Sort the data in descending order by NEI, naming the sorted table `by_nei`.
2. Create another table called `by_nei_pter` that's sorted in descending order by NEI+PTER instead.


_Points:_ 2

In [99]:
by_nei = unemployment.sort('NEI', descending=True)
by_nei_pter = unemployment.sort('NEI+PTER', True)

In [100]:
grader.check("task_10")

task_10 results: All test cases passed!
task_10 - 1 message: ✅ It seems like you have the correct values in by_nei.
task_10 - 2 message: ✅ It seems like you have the correct values in by_nei_pter.

---

### Task 11 📍

For this task:
1. Use `take` to make a table containing the data for the 10 quarters when NEI was greatest.
2. Assign that table to the name `greatest_nei`. Keep in mind that `greatest_nei` should be sorted in descending order of `NEI`.

**Note**: Each row of `unemployment` represents a quarter.


_Points:_ 2

In [104]:
greatest_nei = unemployment.take(np.arange(10))
greatest_nei

DATE       | NEI     | NEI+PTER
1994-01-01 | 9.82912 | 10.9749
1994-04-01 | 9.42543 | 10.6306
1994-07-01 | 9.22842 | 10.359
1994-10-01 | 8.98376 | 10.1199
1995-01-01 | 8.88245 | 10.0086
1995-04-01 | 8.97931 | 10.0942
1995-07-01 | 8.96056 | 10.1052
1995-10-01 | 8.95893 | 10.0813
1996-01-01 | 8.90291 | 9.97698
1996-04-01 | 8.80475 | 9.88223

In [105]:
grader.check("task_11")

task_11 results: All test cases passed!
task_11 - 1 message: ✅ There are 10 rows (quarters) in your table.

---

### Task 12 📍

It's believed that many people became PTER (recall: "Part-Time for Economic Reasons") in the "Great Recession" of 2008-2009.  NEI+PTER is the percentage of people who are unemployed (and counted in the NEI) plus the percentage of people who are PTER.  

Using the original `unemployment` table, create an array containing the percentage of people who were PTER in each quarter.  (The first element of the array should correspond to the first row of `unemployment`, and so on.)

_Points:_ 2

In [108]:
pter = unemployment.column('NEI+PTER')-unemployment.column('NEI')
pter

array([ 1.14582419,  1.20511897,  1.13058532,  1.13616951,  1.12610939,
        1.1148955 ,  1.14462971,  1.12240995,  1.07407416,  1.07748384,
        1.09771919,  1.05719024,  1.03633791,  1.0075395 ,  0.99770422,
        0.96473865,  0.95247726,  0.90691866,  0.88593835,  0.82358655,
        0.84097464,  0.81405539,  0.80697959,  0.76776358,  0.75439202,
        0.75420098,  0.76309996,  0.76412992,  0.76689244,  0.81520797,
        0.87810426,  1.00087012,  0.95177206,  0.94796623,  0.99511177,
        0.97705493,  1.0604883 ,  1.0559854 ,  1.06756208,  1.06831506,
        1.04661712,  1.01607443,  1.01156512,  1.01508214,  0.96581501,
        0.95883018,  1.00737189,  0.91698606,  0.89789276,  0.89124408,
        0.93096016,  0.91680409,  0.91677838,  0.93498812,  0.9857133 ,
        0.95811538,  1.04512065,  1.13518626,  1.29427174,  1.55653413,
        1.84337533,  1.90518743,  1.90376612,  1.90309877,  1.87502276,
        1.867032  ,  1.86025794,  1.85306542,  1.78814381,  1.79

In [109]:
grader.check("task_12")

task_12 results: All test cases passed!
task_12 - 1 message: ✅ pter is an array.
task_12 - 2 message: ✅ You didn't seem to subtract the correct arrays in the wrong order.

---

### Task 13 📍

For this task,
1. Add `pter` as a column to `unemployment` (named `"PTER"`).
2. Make sure the data is sorted by the `'DATE'` values in ascending order (`1994-01-01` should be the first date).
3. Assign this table to the name `unemployment` to update the original table.

**Note**: You do not need to do this all in one line of code.

_Points:_ 2

In [114]:
unemployment = unemployment.with_columns('PTER',pter)
unemployment = unemployment.sort('DATE')
unemployment

DATE       | NEI     | NEI+PTER | PTER
1994-01-01 | 9.82912 | 10.9749  | 1.14582
1994-04-01 | 9.42543 | 10.6306  | 1.20512
1994-07-01 | 9.22842 | 10.359   | 1.13059
1994-10-01 | 8.98376 | 10.1199  | 1.13617
1995-01-01 | 8.88245 | 10.0086  | 1.12611
1995-04-01 | 8.97931 | 10.0942  | 1.1149
1995-07-01 | 8.96056 | 10.1052  | 1.14463
1995-10-01 | 8.95893 | 10.0813  | 1.12241
1996-01-01 | 8.90291 | 9.97698  | 1.07407
1996-04-01 | 8.80475 | 9.88223  | 1.07748
... (110 rows omitted)

In [115]:
unemployment.to_csv('unemployment_.csv')

In [116]:
grader.check("task_13")

task_13 results: All test cases passed!
task_13 - 1 message: ✅ Your table has the correct number of columns.
task_13 - 2 message: ✅ Your table appears to be sorted, in ascending order, by the date values.

---

### Measuring Relative Change

Unemployment measurements vary, but sometimes the way they change can reflect something is wrong with the economy. For example, in the following graph, you should see a sharp increase in unemployment measurements in 2008. 

<img src="./NEI_PTER_over_DATE.png" alt="NEI+PTER values over time" width=400px>

A sudden positive change in unemployment measurements (like what is visualized around 2008, the [Great Recession](https://en.wikipedia.org/wiki/Great_Recession)) would be something to be concerned about. 

There are many benefits of using visualizations to detect such changes, but in this assignment, we want you to start to focus on detecting such changes numerically. 

Run the following code cell to view the first two rows of `unemployment`.

In [117]:
unemployment.show(2)

For the first `'NEI+PTER'` value (January 1, 1994), no change was detected because we don't have an earlier measurement to compare it with. However, in April of the same year, the value went from 10.9749 to 10.6306, which is a change of approximately -0.3444. **The negative change reflects a decrease in unemployment (an increase in employment).** A positive difference would reflect an increase in unemployment (a decrease in employment).

Now, you can't study the amount of change in isolation because a small change during a period of high unemployment could seem dramatic. So, it is better to consider a relative change defined as:

$$\text{relative change} = \frac{\text{new value} - \text{old value}}{\text{old value}}$$

This would mean the relative change in NEI+PTER from January to April, **the first quarter of 1994**, is approximately

$$\frac{10.6306 - 10.9749}{10.9749} \approx	-0.0314.$$

This means there was an approximately 3.14% decrease in unemployment. Without context, it is difficult to understand if -0.0314 is a common negative value or if it is a significantly large negative value. Later in the course, you will learn how to address this concern more systematically. For now, you will just look for the largest negative changes in the data as you complete the next few tasks.


### Task 14 📍

For this task, assign `NEI_PTER_change` to an array with all of the relative changes in NEI+PTER values as we just described.
* The `NEI_PTER_change` array should contain 119 items.
* The first value in `NEI_PTER_change` should be approximately `-0.0314`.
  
**Notes**: 
* We've provided some optional intermediate variables to help you complete this task. You can delete them if you don't end up using them. Just make sure you have `NEI_PTER_change` defined as requested in your code.
* Notice that when you are calculating the relative change values, you'll never divide by the last value in the data set. For now, we recommend using the table method [`exclude`](https://datascience.readthedocs.io/en/master/_autosummary/datascience.tables.Table.exclude.html#datascience.tables.Table.exclude) to handle this. As a hint, `unemployment.exclude(0)` will create a table without the first row of `unemployment`.

_Points:_ 2

In [153]:
NEI_PTER_array = unemployment.column('NEI+PTER')
NEI_PTER_array

array([ 10.974942  ,  10.63055029,  10.35900651,  10.11992939,
        10.00855845,  10.09420951,  10.10519157,  10.08134112,
         9.97698164,   9.88223164,   9.73488492,   9.70224809,
         9.62526017,   9.41571316,   9.24310812,   9.13549589,
         9.03442518,   8.86999505,   8.9216472 ,   8.75865099,
         8.71999786,   8.75343491,   8.63861982,   8.4620134 ,
         8.43357281,   8.42739721,   8.41237045,   8.40252217,
         8.53796948,   8.87319921,   9.17330892,   9.74108786,
         9.80039922,   9.76934566,   9.74537007,   9.72765352,
         9.91252922,  10.00826304,  10.01936281,   9.89609137,
         9.77251611,   9.6694471 ,   9.61485106,   9.59960981,
         9.4905882 ,   9.37175691,   9.3195762 ,   9.28452311,
         9.10574265,   9.04213421,   9.05155693,   8.97961755,
         8.93443143,   9.07792519,   9.14390096,   9.15493675,
         9.37461499,   9.69089358,  10.24812735,  11.0143204 ,
        12.20906974,  12.71757877,  12.67981584,  12.76

In [154]:
first_119_NEI_PTER_array = unemployment.take(np.arange(119)).column("NEI+PTER")
first_119_NEI_PTER_array

array([ 10.974942  ,  10.63055029,  10.35900651,  10.11992939,
        10.00855845,  10.09420951,  10.10519157,  10.08134112,
         9.97698164,   9.88223164,   9.73488492,   9.70224809,
         9.62526017,   9.41571316,   9.24310812,   9.13549589,
         9.03442518,   8.86999505,   8.9216472 ,   8.75865099,
         8.71999786,   8.75343491,   8.63861982,   8.4620134 ,
         8.43357281,   8.42739721,   8.41237045,   8.40252217,
         8.53796948,   8.87319921,   9.17330892,   9.74108786,
         9.80039922,   9.76934566,   9.74537007,   9.72765352,
         9.91252922,  10.00826304,  10.01936281,   9.89609137,
         9.77251611,   9.6694471 ,   9.61485106,   9.59960981,
         9.4905882 ,   9.37175691,   9.3195762 ,   9.28452311,
         9.10574265,   9.04213421,   9.05155693,   8.97961755,
         8.93443143,   9.07792519,   9.14390096,   9.15493675,
         9.37461499,   9.69089358,  10.24812735,  11.0143204 ,
        12.20906974,  12.71757877,  12.67981584,  12.76

In [155]:
NEI_PTER_change = np.diff(NEI_PTER_array)/first_119_NEI_PTER_array
NEI_PTER_change 

array([ -3.13798206e-02,  -2.55437181e-02,  -2.30791548e-02,
        -1.10051111e-02,   8.55778265e-03,   1.08795634e-03,
        -2.36021747e-03,  -1.03517458e-02,  -9.49686071e-03,
        -1.49102679e-02,  -3.35256430e-03,  -7.93506017e-03,
        -2.17705294e-02,  -1.83315948e-02,  -1.16424286e-02,
        -1.10635166e-02,  -1.82003975e-02,   5.82324421e-03,
        -1.82697445e-02,  -4.41313650e-03,   3.83452515e-03,
        -1.31165765e-02,  -2.04438233e-02,  -3.36097126e-03,
        -7.32263891e-04,  -1.78308448e-03,  -1.17068998e-03,
         1.61198391e-02,   3.92634017e-02,   3.38220422e-02,
         6.18946709e-02,   6.08878187e-03,  -3.16860115e-03,
        -2.45416616e-03,  -1.81794453e-03,   1.90051683e-02,
         9.65786041e-03,   1.10906051e-03,  -1.23033211e-02,
        -1.24872793e-02,  -1.05468251e-02,  -5.64624150e-03,
        -1.58517862e-03,  -1.13568788e-02,  -1.25209617e-02,
        -5.56786873e-03,  -3.76123135e-03,  -1.92557507e-02,
        -6.98553031e-03,

In [157]:
NEI_PTER_array = unemployment.column('NEI+PTER')
first_119_NEI_PTER_array = unemployment.take(np.arange(119)).column("NEI+PTER")
NEI_PTER_change = np.diff(NEI_PTER_array)/first_119_NEI_PTER_array
NEI_PTER_change

array([ -3.13798206e-02,  -2.55437181e-02,  -2.30791548e-02,
        -1.10051111e-02,   8.55778265e-03,   1.08795634e-03,
        -2.36021747e-03,  -1.03517458e-02,  -9.49686071e-03,
        -1.49102679e-02,  -3.35256430e-03,  -7.93506017e-03,
        -2.17705294e-02,  -1.83315948e-02,  -1.16424286e-02,
        -1.10635166e-02,  -1.82003975e-02,   5.82324421e-03,
        -1.82697445e-02,  -4.41313650e-03,   3.83452515e-03,
        -1.31165765e-02,  -2.04438233e-02,  -3.36097126e-03,
        -7.32263891e-04,  -1.78308448e-03,  -1.17068998e-03,
         1.61198391e-02,   3.92634017e-02,   3.38220422e-02,
         6.18946709e-02,   6.08878187e-03,  -3.16860115e-03,
        -2.45416616e-03,  -1.81794453e-03,   1.90051683e-02,
         9.65786041e-03,   1.10906051e-03,  -1.23033211e-02,
        -1.24872793e-02,  -1.05468251e-02,  -5.64624150e-03,
        -1.58517862e-03,  -1.13568788e-02,  -1.25209617e-02,
        -5.56786873e-03,  -3.76123135e-03,  -1.92557507e-02,
        -6.98553031e-03,

In [158]:
grader.check("task_14")

task_14 results: All test cases passed!
task_14 - 1 message: ✅ NEI_PTER_change represents an array.
task_14 - 2 message: ✅ NEI_PTER_change has the correct number of items.

Run the following code to add the values in the `NEI+PTER_change` array to the table under the column label of `'NEI+PTER CHANGE'`. (You'll work with `np.append` later in this course.)

In [159]:
# Run this code after completing Task 14
unemployment_with_change = unemployment.with_column('NEI+PTER CHANGE',
                                                    np.append(make_array(0), NEI_PTER_change))
unemployment_with_change

DATE       | NEI     | NEI+PTER | PTER    | NEI+PTER CHANGE
1994-01-01 | 9.82912 | 10.9749  | 1.14582 | 0
1994-04-01 | 9.42543 | 10.6306  | 1.20512 | -0.0313798
1994-07-01 | 9.22842 | 10.359   | 1.13059 | -0.0255437
1994-10-01 | 8.98376 | 10.1199  | 1.13617 | -0.0230792
1995-01-01 | 8.88245 | 10.0086  | 1.12611 | -0.0110051
1995-04-01 | 8.97931 | 10.0942  | 1.1149  | 0.00855778
1995-07-01 | 8.96056 | 10.1052  | 1.14463 | 0.00108796
1995-10-01 | 8.95893 | 10.0813  | 1.12241 | -0.00236022
1996-01-01 | 8.90291 | 9.97698  | 1.07407 | -0.0103517
1996-04-01 | 8.80475 | 9.88223  | 1.07748 | -0.00949686
... (110 rows omitted)

---

### Defining Significance

When dealing with large-scale economic indicators like unemployment, even small percentage changes can represent substantial shifts in the labor market. There are many ways to define when numerical values are significant. Sometimes significance is determined by some, seemingly arbitrary, threshold value.

For the sake of this assignment, let's say that a 10% increase in NEI+PTER measurements reflects a significant (bad) change in unemployment. In other words, **consider any relative change more than 0.1 to be significant.**

### Task 15 📍

For this task, create a table called `significant_changes` that contains only the rows from the unemployment table with a `'NEI+PTER CHANGE`' of more than `0.1`.

_Points:_ 3

In [167]:
significant_changes = unemployment_with_change.where("NEI+PTER CHANGE", are.above(0.1))
significant_changes

DATE       | NEI     | NEI+PTER | PTER    | NEI+PTER CHANGE
2009-01-01 | 10.3657 | 12.2091  | 1.84338 | 0.108472
2020-04-01 | 14.1134 | 16.1138  | 2.00032 | 0.875494

In [166]:
grader.check("task_15")

task_15 results: All test cases passed!
task_15 - 1 message: ✅ significant_changes is a Table.

---

### Task 16 📍🔎

<!-- BEGIN QUESTION -->

What do you notice about the timing of these significant negative changes in NEI+PTER? In your response:
* Highlight any global events that coincide with the dates found.
* Express whether or not the events caused these significant changes in unemployment.

_Points:_ 2

There is a Great Recession (2007-2009), for example, where the unemployment rate has been significantly decreasing, where the economy/market is very low and people cannot afford to purchase necessities for their survival. The depression also coincidences with the significant change in unemployment.

<!-- END QUESTION -->

---

## Submit Your Assignment to Canvas

Follow these steps to submit your homework assignment:

1. **Review the Rubric:** View the rubric on the associated Canvas Assignment page to understand the scoring criteria.
2. **Run the Auto-Grader:** Ensure you have executed the code cell containing the command `grader.check_all()` to run all tests for auto-graded tasks marked with 📍. This command will execute all auto-grader tests sequentially.
3. **Complete Manually Graded Tasks:** Verify that you have responded to all the manually graded tasks marked with 📍🔎.
4. **Save Your Work:** In the notebook's Toolbar, go to `File -> Save Notebook` to save your work and create a checkpoint.
5. **Download the Notebook:** In the notebook's Toolbar, go to `File -> Download IPYNB` to download the notebook (`.ipynb`) file.
6. **Upload to Canvas:** On the Canvas Assignment page, click "Start Assignment" or "New Attempt" to upload the downloaded `.ipynb` file.

---

## Attribution

This content is licensed under the <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License (CC BY-NC-SA 4.0)</a> and derived from the <a href="https://www.data8.org/">Data 8: The Foundations of Data Science</a> offered by the University of California, Berkeley.

<img src="./by-nc-sa.png" width=100px>

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [168]:
grader.check_all()

task_01 results: All test cases passed!
task_01 - 1 message: ✅ fruits is a Table.
task_01 - 2 message: ✅ fruits looks correct!

task_02 results: All test cases passed!
task_02 - 1 message: ✅ inventory is a Table.
task_02 - 2 message: ✅ inventory looks correct.

task_03 results: All test cases passed!
task_03 - 1 message: ✅ all_different is assigned to True or False.

task_04 results: All test cases passed!
task_04 - 1 message: ✅ sales is a Table.
task_04 - 2 message: ✅ sales looks correct.

task_05 results: All test cases passed!
task_05 - 1 message: ✅ total_fruits_sold is an integer.
task_05 - 2 message: ✅ total_fruits_sold is a possible quantity.

task_06 results: All test cases passed!
task_06 - 1 message: ✅ total_revenue is a possible value.
task_06 - 2 message: ✅ total_revenue doesn't just reflect the revenue for grapes.

task_07 results: All test cases passed!
task_07 - 1 message: ✅ remaining_inventory is a Table.
task_07 - 2 message: ✅ The updated grape inventory looks good.

ta